hashcracker/client/lib/client.ex

```diff
@@ -0,0 +1,207 @@
+                   attack: :string, num_workers: :integer,
+                   wordlist_path: :string, mask: :string,
+                   increment: :string, help: :boolean,
+                   client: :string, worker_nodes: :string,
+                   cookie: :string]
+  @options [aliases: @options_aliases, strict: @options_strict]
+
+  @name :crackercli
+
+  def main(args) do
+    {options, _, _} = OptionParser.parse(args, @options)
+    options = options
+    |> Enum.into(%{})
+    |> check_for_help
+    |> validate_required_options
+    |> format_required_options
+    |> validate_required_option_args
+    |> validate_mode_options
+    |> set_defaults
+    |> format_additional_options
+    |> setup_node
+
+    Cracker.crack(options, {@name, options.client_node})
+
+    output()
+  end
+
+  def check_for_help(options) do
+    Map.has_key?(options, :help)
+    |> help_handler(options)
+  end
+  def help_handler(false, options) do
+    options
+  end
+  def help_handler(true, _options) do
+    help()
+  end
+
+  def validate_required_options(options) do
+    required = [:hash, :hash_type, :attack, :num_workers]
+    options
+    |> has_required?(required)
+  end
+
+  def format_required_options(options) do
+    options
+    |> Map.update!(:hash_type, &String.to_atom/1)
+    |> Map.update!(:attack, &String.to_atom/1)
+    |> Map.update!(:hash, fn hash -> Base.decode16!(hash, case: :mixed) end)
+  end
+
+  def validate_required_option_args(options) do
+    [:dictionary, :mask, :brute]
+    |> Enum.any?(fn x -> x == options.attack end)
+    |> validation_handler(options)
+  end
+
+  def validate_mode_options(options=%{attack: :mask}) do
+    has_required?(options, [:mask])
+  end
+  def validate_mode_options(options=%{attack: :dictionary}) do
+    has_required?(options, [:wordlist_path])
+  end
+  def validate_mode_options(options) do
+    options
+  end
+
+  def set_defaults(options) do
+    options
+    |> set_client_node
+    |> set_worker_nodes
+  end
+
+  def set_client_node(options) do
+    case Map.has_key?(options, :client_node) do
+      true ->
+        options
+      false ->
+        {:ok, hostname} = :inet.gethostname
+        Map.put(options, :client_node, "client@#{hostname}")
+    end
+  end
+
+  def set_worker_nodes(options) do
+    case Map.has_key?(options, :worker_nodes) do
+      true ->
+        options
+      false ->
+        Map.put(options, :worker_nodes, options.client_node)
+    end
+  end
+
+  def format_additional_options(options=%{increment: increment}) do
+    [start, stop] = String.split(increment, ":")
+                    |> Enum.map(&String.to_integer/1)
+    options
+    |> Map.drop([:increment])
+    |> Map.merge(%{start: start, stop: stop})
+  end
+  def format_additional_options(options) do
+    options
+    |> Map.update!(:client_node, &String.to_atom/1)
+    |> Map.update!(:worker_nodes, fn nodes ->
+      String.split(nodes, ",") |> Enum.map(&String.to_atom/1)
+    end)
+  end
+
```

Comment:
I get a warning when I compile this because options is assigned only inside a conditional. I kinda agree with it. Perhaps tidier code would be:

~~~ elixir
  def setup_node(options) do
    options = set_cookie_from(options)
    {:ok, _} = Node.start(options.client_node, :shortnames)
    Process.register(self(), @name)
    options
  end

  def set_cookie_from(options = %{ cookie: cookie }) do
    options = %{ options | cookie: String.to_atom!(cookie)}
    Node.set_cookie(options.cookie)
    options
  end
  def set_cookie_from_options(options) do
    options
  end
~~~

No deduction: just fine tuning.